mpr-research-data.py

```diff
@@ -144,51 +107,228 @@ def sliceAndPushToGCP(courseDF, retrieveDF, client, targetBucketName=targetBucke
 
         except Exception as e:
             logging.error(f'Error Message: {e}')
-            logging.error(f'Failed to save Course Data for {outputFilename}.')
+            logging.error(
+                f'Failed to save Course Data for {outputFilename}.')
+            allSliced = False
             continue
 
         try:
             logging.info(f'Saving to GCP: {outputFilename}')
             blob = bucket.blob(outputFilename)
             blob.upload_from_filename(outputFilename)
             os.remove(outputFilename)
+
         except Exception as e:
             logging.error(f'Error Message: {e}')
             logging.error(
                 f'Failed to upload Course Data for {outputFilename} to GCP.')
+            allSaved = False
             continue
 
-    return True
+    return allSliced, allSaved
 
 
-# RETRIEVE KEYS
-# --------------------------------------------------------------------------
+class Config:
```

Comment:
I think some of the config item validation code is repetitive.  I wonder whether some of the validation is similar enough that a generalized version of it could be used to apply to multiple config items.  It may even reduce the amount of code.

I'm thinking (not thoroughly tested code here) something like:

```python
def configFetch(name, default=None, casting=None, validation=None):
  value = os.getenv(name, default)
  if (casting is not None):
    value = casting(value)
  if (validation is not None and not validation(value)):
    raise ValueError(f'Validation error for config item "{name}" value "{value}".')
  return value

# A practical application, I think
self.numberOfMonths = configFetch('NUMBER_OF_MONTHS', self.numberOfMonths, 
  int, lambda x: x > 0)

# Maybe contrived, but an example of how flexible this could be.
# Demonstrates the default and validation may be skipped, and the casting may be complex
self.gcpParams = configFetch('GCP_KEY', casting=lambda x: json.loads(str(x)))
```

Maybe my suggestion could even be a reinvention of the wheel.  I'd suggest looking around on [PyPI](http://pypi.python.org/) to see whether some modules already exist for fetching and validating values from environment variables.

Reply:
I have never used a lambda function like that before! That's extremely powerful. I will implement this in the config setup.

Reply:
I see from your commit (a416e7b) that it seems like for the config handling, you ended up cutting the number of lines of code almost in half.